File: Mobile/lib/service/service.dart
import 'dart:convert';

import 'package:flutter/cupertino.dart';
import 'package:http/http.dart' as http;
import 'package:provider/provider.dart';
import '../provider/cart.dart';
import '../provider/coffee.dart';
import '../provider/coffeedata.dart';
import '../provider/orders.dart';
import '../provider/user.dart';
import '../screens/cart_screen/widget/cart_item.dart';

const String host = 'http://192.168.3.104:12121';

class Service{
  static String author = 'https://sharecode.vn/thanh-vien/truong-nguyen-ngoc-dan-447820.htm';

  static Service service = Service();
  static Service gI(){
    return service;
  }

  //Get provider
  User getUser(BuildContext context){
    return Provider.of<User>(context);
  }

  Future<bool> checkLogin(String username, String password, User user) async {
    String url = '$host/login';
    print(url);
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*", // Required for CORS support to work
      "Content-Type" : "application/json"
    };
    Map<String, String> body = {
      "username" : username,
      "password" : password
    };

    http.Response response =
        await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
    final data = jsonDecode(response.body);
    //print(response.body);
    if(data['isSuccess'] == true){
      user.login(data['token'], data['name'], data['image']);
      return true;
    }
    else{
      return false;
    }
  }

  Future<bool> register(String username, String password, String fullName) async {
    String url = '$host/register';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*", // Required for CORS support to work
      "Content-Type" : "application/json"
    };
    Map<String, String> body = {
      "username" : username,
      "password" : password,
      "fullName" : fullName
    };

    http.Response response =
    await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
    final data = jsonDecode(response.body);
    print(response.body);
    return data['isSuccess'] as bool;
  }

  Future<void> getUserCart(User user, Cart cart) async {
    String url = '$host/get_user_cart';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*",
      "Content-Type" : "application/json"
    };
    Map<String, String> body = {
      "jwtKey" : user.key,
    };

    http.Response response =
    await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
    final data = jsonDecode(response.body);
    final cartArr = data['cart'];
    List<CartItems> cartItems = [];
    for(var cartItem in cartArr){
        dynamic cfData = cartItem['coffe'];
        int quantity = cartItem['quantity'];
        CartItems cI = CartItems(id: cfData['id'], title: cfData['title'], image: cfData['image'], amount: cfData['amount'], descritoin: cfData['description'], quantity: quantity);
        cartItems.add(cI);
    }
    cart.loadItem(cartItems);
  }

   Future<void> getUserOrder(User user, Orders orders) async {
    String url = '$host/get_user_order';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*",
      "Content-Type" : "application/json"
    };
    Map<String, String> body = {
      "jwtKey" : user.key,
    };
    http.Response response =
    await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
    final data = jsonDecode(response.body);
    final orderArr = data['orders'];
    List<OrdersItem> ordersItem = [];

    for(var orderItem in orderArr){
      String dateTimeString = orderItem['date_time'];
      String formattedDateTimeString = dateTimeString.replaceAll('"', ''); // Loại bỏ dấu ngoặc kép
      DateTime dateTime = DateTime.parse(formattedDateTimeString);
      List<CartItem> products = [];
      for(var orderItem in orderItem['order_items']){
          dynamic coffeData = orderItem['coffe'];
          CartItem cartItem = CartItem(image: coffeData['image'], title: coffeData['title'],
              prodcutid: coffeData['id'], id: orderItem["id_order"].toString(),
              amount: coffeData['amount'], qunatity: orderItem['quantity'], description: coffeData['description']);
          products.add(cartItem);
      }
      OrdersItem orderI = OrdersItem(id: orderItem['id'].toString(), amount: orderItem['amount'], products: products, dateTime: dateTime);
      ordersItem.add(orderI);
    }
    orders.loadOrders(ordersItem);
  }

  Future<void> getUserBalance(User user) async {
    String url = '$host/get_user_balance';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*",
      "Content-Type": "application/json"
    };
    Map<String, String> body = {
      "jwtKey": user.key,
    };
    http.Response response =
    await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
    final data = jsonDecode(response.body);
    print(data);
    user.balance =  data['balance'];
  }

  void logout(User user){
    user.signOut();
  }

  Future<void> getCoffees(CofeeData coffeeData, User user) async{
    List<Coffee> coffees = [];
    String url = '$host/get_coffees';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*",
      "Content-Type": "application/json"
    };
    Map<String, String> body = {
      "jwtKey": user.key,
    };
    http.Response response =
    await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
    final data = jsonDecode(response.body);
    for(int i = 0; i < data.length; i ++){
      Coffee coffee = Coffee(id: data[i]['id'], title: data[i]['title'], image: data[i]['image'], amount: data[i]['amount'], description: data[i]['description']);
      coffee.isFavarate = data[i]['isFavorite'] ?? false;
      coffees.add(coffee);
    }
    coffeeData.coffeeList = coffees;
  }

  Future<void> updateFavorite(User user, String idCoffee) async{
    List<Coffee> coffees = [];
    String url = '$host/update_favorite';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*",
      "Content-Type": "application/json"
    };
    Map<String, dynamic> body = {
      "jwtKey": user.key,
      "id_coffe" : idCoffee
    };
    http.Response response =
        await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
  }

  Future<bool> changePassword(User user, String newPassword) async{
    String url = '$host/change_password';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*",
      "Content-Type": "application/json"
    };
    Map<String, String> body = {
      "jwtKey": user.key,
      "newPassword" : newPassword
    };
    http.Response response =
    await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
    final data = jsonDecode(response.body);
    print(data);
    return data['isSuccess'];
  }

  Future<void> addItemToCart(User user, List<dynamic> items) async{
    String url = '$host/add_to_cart';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*",
      "Content-Type": "application/json"
    };
    Map<String, String> body = {
      "jwtKey": user.key,
      "coffees" : json.encode(items)
    };
    await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
  }

  Future<void> removeItemInCart(User user, String idCoffee) async{
    String url = '$host/remove_cart';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*",
      "Content-Type": "application/json"
    };
    Map<String, String> body = {
      "jwtKey": user.key,
      "id_coffe" : idCoffee
    };
    await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
  }

  Future<int> placeOrder(User user) async{
    String url = '$host/place_order';
    Map<String, String> headers = {
      "Access-Control-Allow-Origin": "*",
      "Content-Type": "application/json"
    };
    Map<String, String> body = {
      "jwtKey": user.key
    };
    http.Response response =
    await http.post(Uri.parse(url), headers: headers, body: json.encode(body));
    final data = jsonDecode(response.body);
    return data['isSuccess'];
  }
}

File: Backend/Service.py
import json
from DataSource import DataSource
from Security import Security
from Models import *

sc = Security()

class Service:
    def __init__(self):
        self.ds = DataSource()
    
    #Authenciation
    def check_login(self, username: str, password: str):
        user = self.ds.get_user(username, password)
        if user != None:
            return [user, sc.generate_jwt_key(user.id, user.username, user.image, user.name)]
        return [None, None]
    
    def check_jwt(self, token: str) -> bool:
        return sc.validate_jwt_key(token) != None
    
    def addAccount(self, username: str, password: str, fullname: str) -> bool:
        return self.ds.add_user(username, password, fullname)
    
    def getFavoriteCoffes(self, jwtKey: str) -> list:
        favorites = []
        userInfo = sc.validate_jwt_key(jwtKey)
        if userInfo != None:
            idUser = userInfo["id"]
            query = f"select favorite from User where id = '{idUser}'"
            data = self.ds.sql.execute_query(query)[0]
            favorites = json.loads(data[0])
        return favorites
    
    def removeFavorite(self, jwtKey: str, id_coffe: int) -> bool:
        userInfo = sc.validate_jwt_key(jwtKey)
        if userInfo != None:
            idUser = userInfo["id"]
            favorites = self.getFavoriteCoffes(jwtKey)
            try:
                if id_coffe in favorites:
                    favorites.remove(id_coffe)
                else:
                    favorites.append(id_coffe)
            except:
                pass
            query = f"update User set favorite = '{json.dumps(favorites)}' where id = '{idUser}'"
            self.ds.sql.execute_insert_query(query)
            return True
        return False
    
    def changePassword(self, jwtKey: str, newPassword: str) -> bool:
        userInfo = sc.validate_jwt_key(jwtKey)
        if newPassword.isalnum() == False:
            return False
        
        if userInfo != None:
            idUser = userInfo["id"]
            query = f"update User set password = '{newPassword}' where id = '{idUser}'"
            self.ds.sql.execute_insert_query(query)
            return True
        
        return False
    
    #Product
    def get_coffes(self, jwtKey: str) -> list:
        query = "select * from Coffe"
        coffees = []
        favorites = self.getFavoriteCoffes(jwtKey)
        userInfo = sc.validate_jwt_key(jwtKey)
        if userInfo != None:
            for cf_data in self.ds.sql.execute_query(query):
                cf = Coffe(cf_data)
                cf.isFavorite = cf.id in favorites
                coffees.append(cf.to_dict())
        return coffees
    
    def getCart(self, jwtKey: str) -> list:
        userInfo = sc.validate_jwt_key(jwtKey)
        if userInfo != None:
            idUser = userInfo["id"]
            query = f"select cart from User where id = '{idUser}'"
            cart = list(json.loads(self.ds.sql.execute_query(query)[0][0]))
            return cart
        return []

    def addCoffeToCart(self, jwt, coffees):
        userInfo = sc.validate_jwt_key(jwt)
        if userInfo == None:
            return False
        id_user = userInfo["id"]
        queryUpdate = f"update User set cart = '{(coffees)}' where id = '{id_user}'"
        self.ds.sql.execute_insert_query(queryUpdate)
        return True

    def removeCoffeFromCart(self, id_user, id_coffe):
        query = f"select cart from User where id = '{id_user}'"
        cart = list(json.loads(self.ds.sql.execute_query(query)[0][0]))
        for i in range(len(cart)):
            if cart[i]["id_coffee"] == id_coffe:
                cart.pop(i)
                break
        queryUpdate = f"update User set cart = '{json.dumps(cart)}' where id = '{id_user}'"
        self.ds.sql.execute_insert_query(queryUpdate)
        print("cart", cart)
        pass
    

    def removeItemInCart(self, jwtKey, id_coffe):
        userInfo = sc.validate_jwt_key(jwtKey)
        if userInfo != None:
            idUser = userInfo["id"]
            self.removeCoffeFromCart(idUser, id_coffe)
            return True
        return False
    
    def placeOrder(self, jwtKey: str) -> int:
        userInfo = sc.validate_jwt_key(jwtKey)
        if userInfo != None:
            idUser = userInfo["id"]
            query = f"select cart from User where id = '{idUser}'"
            cart = json.loads(self.ds.sql.execute_query(query)[0][0])
            total = 0
            for cart_data in cart:
                id_coffe = cart_data["id_coffee"]
                quantity = cart_data["quantity"]
                queryCoffe = f"select amount from Coffe where id = '{id_coffe}'"
                amount = self.ds.sql.execute_query(queryCoffe)[0][0]
                total += amount * quantity
            currentMoney = self.getUserBalance(jwtKey)["balance"]
            print("total", total)
            print("currentMoney", currentMoney)
            if total > currentMoney:
                return 0
            queryOrder = f"insert into Orders(id_user, amount) values('{idUser}', '{total}')"
            self.ds.sql.execute_insert_query(queryOrder)
            queryOrderId = f"select max(id) from Orders"
            orderId = self.ds.sql.execute_query(queryOrderId)[0][0]
            for cart_data in cart:
                id_coffe = cart_data["id_coffee"]
                quantity = cart_data["quantity"]
                queryOrderItem = f"insert into OrdersItem(id_order, id_coffe, quantity) values('{orderId}', '{id_coffe}', '{quantity}')"
                self.ds.sql.execute_insert_query(queryOrderItem)
            queryUpdate = f"update User set balance = '{currentMoney - total}',cart = '[]' where id = '{idUser}'"
            self.ds.sql.execute_insert_query(queryUpdate)
            return 1
        return -1

    # User
    def getUserCart(self, jwtKey: str) -> list:
        userInfo = sc.validate_jwt_key(jwtKey)
        if userInfo != None:
            idUser = userInfo["id"]
            query = f"select cart from User where id = '{idUser}'"
            cart = json.loads(self.ds.sql.execute_query(query)[0][0])
            cartItems = []
            for cart_data in cart:
                print("cart_data", cart_data)
                id_coffe = cart_data["id_coffee"]
                quantity = cart_data["quantity"]
                cartItem = CartItem(id_coffe=id_coffe, quantity=quantity, data=self.ds.get_coffe_by_id(id_coffe))
                cartItems.append(cartItem.to_dict())
            return {
                "cart": cartItems,
            }
        else:
            return {
                "error": "Invalid token",
                "cart": []
            }
    
    #   final List<CartItem> products;

    def getUserOrder(self, jwtKey: str) -> list:
        userInfo = sc.validate_jwt_key(jwtKey)
        if userInfo != None:
            idUser = userInfo["id"]
            query = f"select * from Orders where id_user = '{idUser}'"
            orders = []
            for order_data in self.ds.sql.execute_query(query):
                order = Order(order_data)
                queryOrderItem = f"select * from OrdersItem oi join Coffe cf on cf.id = oi.id_coffe where id_order = '{order.id}'"
                for order_data in self.ds.sql.execute_query(queryOrderItem):
                    orderItem = OrderItem(order_data)
                    order.add_order_item(orderItem)
                orders.append(order.to_dict())
            return {
                "orders": orders
            }
        else:
            return {
                "error": "Invalid token",
                "orders": []
            }
    
    def getUserBalance(self, jwtKey: str) -> dict:
        userInfo = sc.validate_jwt_key(jwtKey)
        if userInfo != None:
            idUser = userInfo["id"]
            query = f"select balance from User where id = '{idUser}'"
            balance = self.ds.sql.execute_query(query)[0][0]
            return {
                "balance": balance
            }
        else:
            return {
                "error": "Invalid token",
                "balance": 0.0
            }
    